## Final Report

In [1]:
import pandas as pd
import numpy as np
import wrangle as w

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_rows', 1000)

In [2]:
#get the data
df = w.acquire_anonymized_curriculum_access_data()

In [3]:
#look at the data
df

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
900218  2021-04-21  16:41:51   
900219  2021-04-21  16:42:02   
900220  2021-04-21  16:42:09   
900221  2021-04-21  16:44:37   
900222  2021-04-21  16:44:39   

                                                    page  user_id  cohort_id  \
0                                                      /        1        8.0   
1                                                java-ii        1        8.0   
2                    java-ii/object-oriented-programming        1        8.0   
3                     slides/object_oriented_programming        1        8.0   
4                              javascript-i/conditionals        2       22.0   
...                                                  ...      ...        ...   
900218                              jquery/personal-site       64       28.0   
900219                                 jquery/mapbox-api       64       28.0   
900220                           jquery/ajax/weather-map       64       28.0   
900221  anomaly-detection/discrete-probabilistic-methods      744       28.0   
900222                                 jquery/mapbox-api       64       28.0   

                   ip     name  start_date    end_date  program_id  
0        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
1        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
2        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
3        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
4        97.105.19.61    Teddy  2018-01-08  2018-05-17         2.0  
...               ...      ...         ...         ...         ...  
900218  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900219  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900220  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900221  24.160.137.86    Staff  2014-02-04  2014-02-04         2.0  
900222  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  

[900223 rows x 10 columns]

In [4]:
#look at the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900223 entries, 0 to 900222
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   date        900223 non-null  object 
 1   time        900223 non-null  object 
 2   page        900222 non-null  object 
 3   user_id     900223 non-null  int64  
 4   cohort_id   847330 non-null  float64
 5   ip          900223 non-null  object 
 6   name        847330 non-null  object 
 7   start_date  847330 non-null  object 
 8   end_date    847330 non-null  object 
 9   program_id  847330 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 68.7+ MB


In [5]:
# clean the data
df = w.clean_the_data(df)

## 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

**Findings:**
- Data Science accessed classification/overview 1785 times.
- Front End accessed content/html-css 2 times.
- Java accessed javascript-i 17457 times.
- PHP accessed javascript-i 736 times.

In [6]:
#Make df for each program
ds=df[df.program_id=='Data Science']
fe=df[df.program_id=='Front End']
jv=df[df.program_id=='Java']
php=df[df.program_id=='PHP']

In [7]:
#Look at what page was most viewed for data science
ds.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                                                                                                                                                                                                                                                                                                         8358
search/search_index.json                                                                                                                                                                                                                                                                                                  2203
classification/overview                                                                                                                                                                                                                                                                                                   1785
1-fundamentals/modern-data-scientist.j

In [8]:
#Look at what page was most viewed for front end
fe.groupby(['page'])['page'].count().sort_values(ascending=False)

page
content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

In [9]:
#Look at what page was most viewed for java
jv.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                                                      35814
javascript-i                                                           17457
toc                                                                    17428
search/search_index.json                                               15212
java-iii                                                               12683
                                                                       ...  
content/mysql/relationships/sub-queries.html                               1
content/mysql/sample-database/gitbook/images/favicon.ico                   1
content/php_iii/classes-and-objects-i/constructors-destructors.html        1
content/php_iv/classes-and-objects-ii/late-static-binding.html             1
wp-login                                                                   1
Name: page, Length: 1913, dtype: int64

In [10]:
#Look at what page was most viewed for php
php.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                                                                1681
index.html                                                                       1011
javascript-i                                                                      736
html-css                                                                          542
spring                                                                            501
java-iii                                                                          479
java-ii                                                                           454
java-i                                                                            444
javascript-ii                                                                     429
appendix                                                                          409
jquery                                                                            344
mysql                                            

## 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

**Findings:**
- Cohorts **Kings, Quincy, and Joshua** all look at PHP in their top 10 most visited lessons. They also have differently marked pathways to their content which could suggest that they made significant changes before or after these cohorts.
- Regression lesson is most viewed for **Curie** with about 600 clicks where as it barely shows up one other time in the top 6 lessons for other data science cohorts

In [11]:
#make df of just those in Data Science 
DS = df[df.program_id == 'Data Science']

#make df of just those in WebDev 
WebDev = df[(df.program_id == 'PHP') | (df.program_id == 'Java') | (df.program_id == 'Front End') ]

In [12]:
DS_Cohorts = DS.groupby('page')['name'].count().sort_values()

In [13]:
DS_Cohorts

page
%20https://github.com/RaulCPena                                                                                                                                                                                                                                                                                              1
appendix/open_data/www.census.gov                                                                                                                                                                                                                                                                                            1
appendix/open_data/www.flickr.com/services/api                                                                                                                                                                                                                                                                               1
appendix/open_data/www.openstreetmap.o

In [14]:
WebDev_Cohorts = WebDev.groupby('page')['name'].count().sort_values()

In [15]:
WebDev_Cohorts

page
'                               1
cli-02-listing-files            1
cli-03-file-paths               1
cli-07-more-topics              1
codeup-exercises                1
                            ...  
java-iii                    13162
search/search_index.json    15331
toc                         17580
javascript-i                18193
/                           37496
Name: name, Length: 2054, dtype: int64

In [16]:
WebDevDict = {}
for x in WebDev.name.unique():
    WebDevDict[x] = pd.DataFrame(df[df.name == x])

In [17]:
DSDict = {}
for x in DS.name.unique():
    DSDict[x] = pd.DataFrame(df[df.name == x])

In [18]:
#cohorts Kings, quincy, and joshua all look at PHP in their top 10 most visited lessons. they also have differently marked pathways to their content which could suggest that they made significant changes before or after these cohorts.
for x in WebDevDict:
    display(x, WebDevDict[x].groupby('page')['name'].count().sort_values(ascending=False).head(10))


'Hampton'

page
/                    210
java-iii              57
appendix              55
java-i                46
java-ii               46
spring                43
javascript-i          36
javascript-ii         31
java-iii/servlets     30
mysql                 30
Name: name, dtype: int64

'Teddy'

page
/                           1828
spring                       744
java-iii                     732
mysql                        638
javascript-i                 615
mkdocs/search_index.json     595
java-i                       522
jquery                       512
java-ii                      509
appendix                     505
Name: name, dtype: int64

'Sequoia'

page
/                                         630
spring                                    364
spring/fundamentals/views                 207
javascript-i                              202
spring/fundamentals/controllers           190
spring/fundamentals/repositories          172
appendix                                  170
mkdocs/search_index.json                  168
html-css                                  168
spring/fundamentals/form-model-binding    156
Name: name, dtype: int64

'Arches'

page
/                626
javascript-i     294
html-css         215
javascript-ii    204
spring           192
java-ii          167
java-iii         153
jquery           133
appendix         133
java-i           132
Name: name, dtype: int64

'Niagara'

page
spring           65
index.html       63
/                37
java-iii         29
mysql            24
javascript-i     23
javascript-ii    21
appendix         20
java-i           18
java-ii          17
Name: name, dtype: int64

'Pinnacles'

page
/                    149
javascript-i          69
javascript-ii         58
spring                56
html-css              56
java-i                54
javascript-ii/npm     51
java-ii               50
java-iii              45
mysql                 39
Name: name, dtype: int64

'Quincy'

page
/                                                   151
content/laravel/intro                                38
content/html-css                                     33
content/laravel/intro/environment.html               31
content/laravel/intro/application-structure.html     20
content/php_iii                                      19
content/mysql/intro-to-mysql                         19
javascript-i/conditionals                            16
content/php_iii/php-with-html                        15
content/appendix                                     14
Name: name, dtype: int64

'Kings'

page
/                                                   219
index.html                                           84
content/laravel/intro                                83
content/laravel/intro/application-structure.html     63
content/laravel/intro/gitbook/images/favicon.ico     56
content/php_iii                                      47
content/html-css                                     45
content/php_iv                                       42
content/php_ii                                       40
content/laravel/quickstart/db-migrations.html        40
Name: name, dtype: int64

'Lassen'

page
index.html                          877
javascript-i                        233
java-iii                            224
spring                              222
html-css                            174
appendix                            170
java-ii                             160
java-i                              138
/                                   125
spring/fundamentals/repositories    115
Name: name, dtype: int64

'Mammoth'

page
/                                       33
java-i                                  19
java-iii                                13
spring                                  12
spring/fundamentals/controllers         12
spring/fundamentals/views               11
java-ii                                 10
javascript-ii                           10
java-iii/deployment-and-dependencies     9
java-iii/jsp-and-jstl                    9
Name: name, dtype: int64

'Glacier'

page
/                                                                            51
content/html-css                                                             29
javascript-i/introduction/working-with-data-types-operators-and-variables    24
html-css                                                                     21
javascript-i/functions                                                       20
content/html-css/introduction.html                                           20
javascript-i/introduction/variables                                          19
javascript-i/introduction/operators                                          18
javascript-i/conditionals                                                    16
javascript-i/javascript-with-html                                            14
Name: name, dtype: int64

'Denali'

page
/                            1
mkdocs/search_index.json     1
prework/databases            1
prework/versioning/github    1
Name: name, dtype: int64

'Joshua'

page
content/html-css                           19
content/html-css/introduction.html         14
content/javascript                         12
content/mysql/intro-to-mysql               10
content/html-css/elements.html              9
/                                           8
index.html                                  8
content/html-css/forms.html                 7
content/php_iv                              7
content/javascript/primitive-types.html     7
Name: name, dtype: int64

'Olympic'

page
/                                                                            249
javascript-i                                                                 128
java-i                                                                        76
jquery                                                                        71
html-css                                                                      69
java-i/methods                                                                69
javascript-i/introduction/working-with-data-types-operators-and-variables     65
javascript-i/javascript-with-html                                             63
javascript-i/bom-and-dom/bom                                                  60
java-i/console-io                                                             60
Name: name, dtype: int64

'Ulysses'

page
/                           1641
mkdocs/search_index.json     721
html-css                     562
javascript-i                 526
java-ii                      425
java-iii                     412
spring                       409
java-i                       377
mysql                        367
jquery                       364
Name: name, dtype: int64

'Badlands'

page
/                                                     17
content/php_ii/command-line                            6
content/php_i                                          5
content/php_ii/control-structures-i                    5
content/php_ii/command-line/io-streams.html            5
content/php_ii                                         4
content/php_i/intro-via-interactive-shell              3
toc                                                    2
content/php_i/intro-to-operators                       2
content/php_ii/command-line/passing-arguments.html     2
Name: name, dtype: int64

'Apollo'

page
content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: name, dtype: int64

'Ike'

page
/                        19
html-css                 12
html-css/elements        11
html-css/introduction    10
content/html-css          9
html-css/forms            8
javascript-i              7
index.html                6
html-css/css-i            6
toc                       6
Name: name, dtype: int64

'Voyageurs'

page
/                2101
javascript-i      884
java-iii          770
java-ii           756
mysql             663
spring            650
java-i            641
javascript-ii     584
jquery            583
html-css          528
Name: name, dtype: int64

'Wrangell'

page
/                                        1132
toc                                       991
javascript-i                              560
search/search_index.json                  504
html-css                                  411
java-i                                    406
jquery                                    377
java-iii                                  337
java-ii                                   325
html-css/css-ii/bootstrap-grid-system     313
Name: name, dtype: int64

'Xanadu'

page
/                           924
javascript-i                722
html-css                    597
search/search_index.json    577
jquery                      544
mysql                       509
java-ii                     502
java-iii                    497
javascript-ii               496
java-i                      484
Name: name, dtype: int64

'Franklin'

page
javascript-ii/es6      10
/                       4
java-i                  3
java-iii/user-input     3
java-iii                3
java-ii                 3
spring                  2
mysql                   2
jquery                  2
javascript-ii           2
Name: name, dtype: int64

'Yosemite'

page
/                                  981
toc                                697
javascript-i                       452
html-css                           384
spring                             381
search/search_index.json           361
java-iii                           351
javascript-ii                      337
mysql                              331
javascript-ii/map-filter-reduce    299
Name: name, dtype: int64

'Staff'

page
/                           6340
javascript-i                1817
spring                      1403
java-iii                    1393
search/search_index.json    1349
appendix                    1288
html-css                    1284
java-ii                     1260
jquery/ajax/weather-map     1050
javascript-ii                993
Name: name, dtype: int64

'Zion'

page
/                           1798
toc                         1465
javascript-i                 897
java-iii                     753
search/search_index.json     700
html-css                     675
spring                       672
javascript-ii                647
java-ii                      624
java-i                       605
Name: name, dtype: int64

'Andromeda'

page
/               1174
toc              638
javascript-i     516
spring           487
java-iii         433
java-ii          398
html-css         375
mysql            355
java-i           338
jquery           338
Name: name, dtype: int64

'Betelgeuse'

page
/                           955
search/search_index.json    761
javascript-i                717
toc                         610
jquery                      542
html-css/elements           495
java-i                      482
java-ii                     481
java-iii                    472
html-css                    457
Name: name, dtype: int64

'Ceres'

page
/                           1653
search/search_index.json    1380
javascript-i                1003
toc                          911
html-css                     766
java-iii                     682
java-ii                      681
jquery                       637
mysql                        622
spring                       562
Name: name, dtype: int64

'Deimos'

page
/                           1467
javascript-i                 896
html-css                     773
toc                          750
search/search_index.json     746
jquery                       563
spring                       533
java-iii                     523
mysql                        517
java-ii                      511
Name: name, dtype: int64

'Europa'

page
/                           1295
toc                          964
search/search_index.json     742
javascript-i                 506
java-iii                     451
html-css/elements            430
java-ii                      375
mysql                        372
jquery                       370
html-css                     353
Name: name, dtype: int64

'Fortuna'

page
/                           2038
toc                         1293
search/search_index.json    1020
java-iii                     786
javascript-i                 785
java-ii                      657
spring                       636
mysql                        591
html-css                     585
java-i                       555
Name: name, dtype: int64

'Apex'

page
search/search_index.json    1497
/                           1346
toc                          810
html-css                     738
javascript-i                 669
java-ii                      666
java-iii                     663
spring                       600
java-i                       590
mysql                        483
Name: name, dtype: int64

'Ganymede'

page
/                           1727
search/search_index.json    1095
toc                          822
javascript-i                 760
java-iii                     651
java-ii                      641
appendix                     561
jquery                       536
javascript-ii                525
java-i                       515
Name: name, dtype: int64

'Everglades'

page
/    1
Name: name, dtype: int64

'Hyperion'

page
/                           1365
toc                         1022
javascript-i                 910
java-iii                     701
java-ii                      688
search/search_index.json     670
mysql                        553
spring                       526
jquery                       497
java-i                       489
Name: name, dtype: int64

'Bash'

page
/                            772
search/search_index.json     660
toc                          404
javascript-i                 374
java-iii                     324
spring                       311
java-ii                      268
html-css                     265
jquery                       262
javascript-i/introduction    256
Name: name, dtype: int64

'Jupiter'

page
toc                         1866
/                           1696
search/search_index.json     998
javascript-i                 926
java-iii                     795
html-css                     784
java-ii                      755
spring                       621
mysql                        564
java-i                       503
Name: name, dtype: int64

'Kalypso'

page
/                           939
javascript-i                631
toc                         495
html-css                    495
jquery                      470
search/search_index.json    467
java-iii                    456
javascript-ii               440
java-ii                     417
java-i                      408
Name: name, dtype: int64

'Luna'

page
javascript-i                                                                 676
/                                                                            491
search/search_index.json                                                     405
toc                                                                          373
javascript-i/introduction/working-with-data-types-operators-and-variables    315
html-css                                                                     307
jquery                                                                       301
java-iii                                                                     259
java-ii                                                                      255
javascript-ii                                                                247
Name: name, dtype: int64

'Marco'

page
toc                                                                          910
javascript-i                                                                 907
/                                                                            699
javascript-i/introduction/working-with-data-types-operators-and-variables    505
html-css                                                                     479
javascript-i/javascript-with-html                                            411
jquery                                                                       405
javascript-i/functions                                                       382
javascript-i/loops                                                           312
javascript-i/conditionals                                                    292
Name: name, dtype: int64

'Neptune'

page
/                                                                            472
toc                                                                          451
javascript-i/introduction/working-with-data-types-operators-and-variables    329
html-css                                                                     313
javascript-i/javascript-with-html                                            311
javascript-i                                                                 272
html-css/css-ii/bootstrap-grid-system                                        256
html-css/css-ii/bootstrap-introduction                                       227
javascript-i/loops                                                           226
javascript-i/functions                                                       217
Name: name, dtype: int64

'Oberon'

page
javascript-i/introduction/operators                                          179
javascript-i/introduction/primitive-types                                    177
javascript-i/functions                                                       171
javascript-i/introduction/working-with-data-types-operators-and-variables    146
javascript-i/conditionals                                                    144
javascript-i                                                                 136
/                                                                            131
javascript-i/javascript-with-html                                            125
toc                                                                          101
javascript-i/introduction/variables                                           95
Name: name, dtype: int64

In [19]:
# Regression lesson in most viewed for Curie  with ab 600 clicks where as it barely shows up one other time in the top 6 lessons for other data science cohorts
for x in DSDict:
    display(x, DSDict[x].groupby('page')['name'].count().sort_values(ascending=False).head(7))


'Bayes'

page
/                                                    1967
1-fundamentals/modern-data-scientist.jpg              650
1-fundamentals/AI-ML-DL-timeline.jpg                  648
1-fundamentals/1.1-intro-to-data-science              640
search/search_index.json                              588
6-regression/1-overview                               521
10-anomaly-detection/AnomalyDetectionCartoon.jpeg     387
Name: name, dtype: int64

'Curie'

page
/                                           1712
6-regression/1-overview                      595
search/search_index.json                     538
1-fundamentals/modern-data-scientist.jpg     467
1-fundamentals/AI-ML-DL-timeline.jpg         465
1-fundamentals/1.1-intro-to-data-science     461
3-sql/1-mysql-overview                       441
Name: name, dtype: int64

'Darden'

page
/                                                 2980
classification/overview                           1109
classification/scale_features_or_not.svg           943
sql/mysql-overview                                 774
search/search_index.json                           664
anomaly-detection/AnomalyDetectionCartoon.jpeg     612
anomaly-detection/overview                         592
Name: name, dtype: int64

'Easley'

page
/                                                                1115
classification/scale_features_or_not.svg                          463
classification/overview                                           445
classification/classical_programming_vs_machine_learning.jpeg     432
fundamentals/AI-ML-DL-timeline.jpg                                381
fundamentals/modern-data-scientist.jpg                            379
fundamentals/intro-to-data-science                                372
Name: name, dtype: int64

'Florence'

page
fundamentals/modern-data-scientist.jpg    627
fundamentals/AI-ML-DL-timeline.jpg        624
fundamentals/intro-to-data-science        615
/                                         584
python/data-types-and-variables           258
fundamentals/DataToAction_v2.jpg          208
sql/mysql-overview                        203
Name: name, dtype: int64

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

**Findings:**
- There are 233 users that didn't access the curriculum during their cohort timeframe, but some of those don't have a cohort listed (78)
- I defined "hardly active" as those students who accessed less than or equal to 10% of the average usage (pages viewed)
- 54 users accessed less than or equal to 10% of the average number of pages viewed per user.
- All but one user was from the Java program, user_id 679 was data science(assumed quit the course)
- **Oberon** and **Sequoia** cohorts only had students access the curriculum up through the first month of the course. ***This might mean that these cohorts used a different curriculum platform or there is something wrong with the start and end dates in the db***
- user_ids 24, 41, 371, 741, 56, 57, 45, 36, 250, 65, 52, 33, 253, 278 only started accessing the curriculum when the cohort was almost finished 
- user_ids 879, 832,  388, 572, 772, 529, 543, 46, 13, 702, 806 accessed the curriculum on the first day (or first week) and then stopped. ***I assume these are students who dropped out.***



Below is the code that helped me discover these finidings.

In [20]:
#only look at students
students = df[df['name'] != 'Staff']

In [21]:
#find those who accessed the curriculum during their class
active_student_access = students.query('start_date < date < end_date')

In [22]:
usage = active_student_access.user_id.value_counts().sort_values()

In [23]:
#I defined "hardly active" as those students who accessed less than or equal to 10% of the average usage (pages viewed)
usage.mean() * .10

89.00359116022099

In [24]:
#how many accessed less than 10% of the average -- 54
usage[usage <= 89]

879     1
679     3
832     3
278     4
956     5
812     7
388     8
572    12
889    16
968    16
487    17
963    17
961    18
772    20
49     25
64     25
976    25
24     26
41     28
972    31
973    32
967    33
529    35
959    38
971    41
964    42
371    42
981    42
960    45
741    47
978    49
965    51
56     52
962    52
543    53
46     54
57     62
969    62
858    63
45     65
36     65
975    66
250    67
65     73
52     75
977    76
974    76
13     79
33     79
955    80
702    85
806    88
954    88
253    89
Name: user_id, dtype: int64

In [25]:
#test oddities of those who didn't view many times during cohort
active_student_access[active_student_access.user_id == 24]

date      time                                    page  \
date                                                                      
2018-01-26 2018-01-26  11:24:17                                appendix   
2018-01-26 2018-01-26  11:24:29              appendix/capstone-workbook   
2018-01-29 2018-01-29  11:25:49                                       /   
2018-01-29 2018-01-29  11:25:54                                  spring   
2018-01-29 2018-01-29  11:25:59               spring/fundamentals/views   
2018-01-29 2018-01-29  13:28:23                mkdocs/search_index.json   
2018-01-29 2018-01-29  13:28:33                                html-css   
2018-01-29 2018-01-29  13:28:41  html-css/css-ii/bootstrap-introduction   
2018-01-29 2018-01-29  13:29:50                                       /   
2018-01-29 2018-01-29  13:29:53                                html-css   
2018-01-29 2018-01-29  13:29:56  html-css/css-ii/bootstrap-introduction   
2018-01-30 2018-01-30  09:30:44                            javascript-i   
2018-01-30 2018-01-30  09:30:49            javascript-i/google-maps-api   
2018-01-30 2018-01-30  09:32:13                mkdocs/search_index.json   
2018-01-30 2018-01-30  09:32:21                mkdocs/search_index.json   
2018-02-01 2018-02-01  10:21:40                                  spring   
2018-02-01 2018-02-01  10:21:45         spring/fundamentals/controllers   
2018-02-01 2018-02-01  10:22:10                                java-iii   
2018-02-01 2018-02-01  10:22:16    java-iii/deployment-and-dependencies   
2018-02-01 2018-02-01  10:22:38      slides/deployment_and_dependencies   
2018-02-01 2018-02-01  10:22:46      slides/deployment_and_dependencies   
2018-02-01 2018-02-01  10:22:51                       java-iii/servlets   
2018-02-01 2018-02-01  10:22:56                         slides/servlets   
2018-02-01 2018-02-01  10:22:59                         slides/servlets   
2018-02-07 2018-02-07  11:07:28                                appendix   
2018-02-07 2018-02-07  11:07:31              appendix/capstone-workbook   

            user_id  cohort_id            ip     name start_date   end_date  \
date                                                                          
2018-01-26       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-26       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24 

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

## 5. At some point in 2019, the ability for students and alumni to access both curriculums (web dev to ds, ds to web dev) should have been shut off. Do you see any evidence of that happening? Did it happen before?

## 6. What topics are grads continuing to reference after graduation and into their jobs (for each program)?

<div class="alert alert-info">

    
# Q#6 
# ANSWER:

## 1. Data Science
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|sql/mysql-overview|275|38|
|classification/overview|266|39|
|anomaly-detection/overview|191|31|
|fundamentals/intro-to-data-science|184|36|

## 2. PHP

|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|javascript-i	|736|	35|
|html-css	|542|	40|
|spring	|501|	25|
|java-iii	|479|	20|
    
## 3. Java
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|javascript-i	|4229|	300|
|spring	|3760|	306|
|html-css	|3136|	296|
|java-iii	|3058|	273|
    
## 4. Front End    
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|content/html-css	|2	|1|
|content/html-css/introduction.html	|1	|1|    

# Grad Users

In [26]:
# Create DataFrame for users that have accessed the curriculum past there graduation date
grad_users = df[df.date > df.end_date]

In [27]:
grad_users.T

date                 2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   09:55:03             09:56:02   
page                          /              java-ii   
user_id                       1                    1   
cohort_id                   8.0                  8.0   
ip                 97.105.19.61         97.105.19.61   
name                    Hampton              Hampton   
start_date  2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date    2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                  PHP                  PHP   

date                                 2018-01-26  \
date                        2018-01-26 00:00:00   
time                                   09:56:05   
page        java-ii/object-oriented-programming   
user_id                                       1   
cohort_id                                   8.0   
ip                                 97.105.19.61   
name                                    Hampton   
start_date                  2015-09-22 00:00:00   
end_date                    2016-02-06 00:00:00   
program_id                                  PHP   

date                                2018-01-26           2018-01-26  \
date                       2018-01-26 00:00:00  2018-01-26 00:00:00   
time                                  09:56:06             10:14:47   
page        slides/object_oriented_programming                    /   
user_id                                      1                   11   
cohort_id                                  8.0                  1.0   
ip                                97.105.19.61         97.105.19.61   
name                                   Hampton               Arches   
start_date                 2015-09-22 00:00:00  2014-02-04 00:00:00   
end_date                   2016-02-06 00:00:00  2014-04-22 00:00:00   
program_id                                 PHP                  PHP   

date                 2018-01-26              2018-01-26  \
date        2018-01-26 00:00:00     2018-01-26 00:00:00   
time                   10:14:51                10:15:02   
page               javascript-i  javascript-i/functions   
user_id                      11                      11   
cohort_id                   1.0                     1.0   
ip                 97.105.19.61            97.105.19.61   
name                     Arches                  Arches   
start_date  2014-02-04 00:00:00     2014-02-04 00:00:00   
end_date    2014-04-22 00:00:00     2014-04-22 00:00:00   
program_id                  PHP                     PHP   

date                    2018-01-26           2018-01-26           2018-01-26  \
date           2018-01-26 00:00:00  2018-01-26 00:00:00  2018-01-26 00:00:00   
time                      10:40:15             10:59:02             10:59:06   
page        javascript-i/functions           index.html               java-i   
user_id                          1                   21                   21   
cohort_id                      8.0                 16.0                 16.0   
ip                    97.105.19.61        71.137.146.71        71.137.146.71   
name                       Hampton              Niagara              Niagara   
start_date     2015-09-22 00:00:00  2016-10-26 00:00:00  2016-10-26 00:00:00   
end_date       2016-02-06 00:00:00  2017-03-09 00:00:00  2017-03-09 00:00:00   
program_id                     PHP                 Java                 Java   

date        ...                 2021-04-21           2021-04-21  \
date        ...        2021-04-21 00:00:00  2021-04-21 00:00:00   
time        ...                   16:07:17             16:30:30   
page        ...  javascript-ii/RESTful-api             appendix   
user_id     ...                         64                  771   
cohort_id   ...                       28.0                 62.0   
ip          ...              71.150.217.33      172.125.226.175   
name        ...                      St

In [28]:
# LOOK AT GRAD USERS
grad_program_views = grad_users.groupby(['program_id'])['page'].agg(['count', 'nunique'])
grad_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          157786     1656
PHP            30548      710
Data Science   11495      437
Front End          5        4

In [29]:
# LOOK AT WHOLE DF
df_program_views = df.groupby(['program_id'])['page'].agg(['count', 'nunique'])
df_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          713365     1913
Data Science  103411      682
nan            52893     1112
PHP            30548      710
Front End          5        4

In [30]:
grad_users.groupby('program_id')['page'].value_counts()

program_id    page                                    
Data Science  /                                           1436
              search/search_index.json                     493
              sql/mysql-overview                           275
              classification/overview                      266
              classification/scale_features_or_not.svg     219
                                                          ... 
PHP           students                                       1
              students/468/notes                             1
              students/units/75/sub_units/268                1
              teams/13                                       1
              uploads/58a217a705bde.jpg                      1
Name: page, Length: 2807, dtype: int64

# Data Science

In [31]:
# Create DataFrame for Data Science users that have accessed the curriculum past there graduation date
ds_grad_users = grad_users[grad_users.program_id == 'Data Science']

In [32]:
# Count how many times each page was accessed
ds_grad_users.page.value_counts().head(10)

/                                                 1436
search/search_index.json                           493
sql/mysql-overview                                 275
classification/overview                            266
classification/scale_features_or_not.svg           219
anomaly-detection/AnomalyDetectionCartoon.jpeg     193
anomaly-detection/overview                         191
fundamentals/AI-ML-DL-timeline.jpg                 189
fundamentals/modern-data-scientist.jpg             187
fundamentals/intro-to-data-science                 184
Name: page, dtype: int64

In [33]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

## Check the min/max Graduation Dates.

In [34]:
# Create Java for DS users
DS_users = df[df.program_id == 'Data Science']

In [35]:
# show max end date for DS users
DS_users.end_date.min()

Timestamp('2020-01-30 00:00:00')

In [36]:
# show max end date for DS users
DS_users.end_date.max()

Timestamp('2021-09-03 00:00:00')

In [37]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# PHP

In [38]:
# Create DataFrame for PHP users that have accessed the curriculum past there graduation date
php_grad_users = grad_users[grad_users.program_id == 'PHP']

In [39]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

## Why are Curriculum access counts the same for PHP during and after graduation?

In [40]:
# Create DataFrame for PHP users
php_users = df[df.program_id == 'PHP']

In [41]:
# show max end date for Front End users
php_users.end_date.max()

Timestamp('2017-09-22 00:00:00')

In [42]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [43]:
# Count how many times each page was accessed
php_users.page.value_counts().head(10)

/                1681
index.html       1011
javascript-i      736
html-css          542
spring            501
java-iii          479
java-ii           454
java-i            444
javascript-ii     429
appendix          409
Name: page, dtype: int64

## ANSWER: Because the last of the PHP users graduated in 2017, before the start date of our data access collection which begins in 2018.

# JAVA

In [44]:
# Create DataFrame for Java users that have accessed the curriculum past there graduation date
java_grad_users = grad_users[grad_users.program_id == 'Java']

In [45]:
# Count how many times each page was accessed
java_grad_users.page.value_counts().head(10)

/                           12406
javascript-i                 4229
spring                       3760
search/search_index.json     3562
html-css                     3136
java-iii                     3058
java-ii                      2985
java-i                       2679
appendix                     2662
javascript-ii                2549
Name: page, dtype: int64

In [46]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

## Check the min/max Graduation Dates.

In [47]:
# Create Java for PHP users
java_users = df[df.program_id == 'Java']

In [48]:
# show max end date for Java users
java_users.end_date.min()

Timestamp('2014-02-04 00:00:00')

In [49]:
# show max end date for Java users
java_users.end_date.max()

Timestamp('2021-10-01 00:00:00')

In [50]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# Front End

In [51]:
# Create DataFrame for Front End users that have accessed the curriculum past there graduation date
fe_grad_users = grad_users[grad_users.program_id == 'Front End']

In [52]:
# Count how many times each page was accessed
fe_grad_users.page.value_counts().head(10)

content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

In [53]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

In [54]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

In [55]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

In [56]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

In [57]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

## 7. Which lessons are least accessed?

**Findings:**
- Data Science accessed pages with the prefix appendix the least.
- Front End accessed pages with the prefix content/html-css the least.
- Java accessed pages with the prefix cli- the least.
- PHP accessed pages with the prefix content/appendix/javascript/functions/ the least.

In [58]:
#Look at what page was least viewed for data science
ds.groupby(['page'])['page'].count().sort_values()

page
%20https://github.com/RaulCPena                                                                                                                                                                                                                                                                                              1
appendix/open_data/www.census.gov                                                                                                                                                                                                                                                                                            1
appendix/open_data/www.flickr.com/services/api                                                                                                                                                                                                                                                                               1
appendix/open_data/www.openstreetmap.o

In [59]:
#Look at what page was least viewed for front end
fe.groupby(['page'])['page'].count().sort_values()

page
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
content/html-css                               2
Name: page, dtype: int64

In [60]:
#Look at what page was least viewed for java
jv.groupby(['page'])['page'].count().sort_values()

page
'                                            1
cheatsheet.tex                               1
classification/user-defined-functions        1
cli-02-listing-files                         1
cli-03-file-paths                            1
                                         ...  
java-iii                                 12683
search/search_index.json                 15212
toc                                      17428
javascript-i                             17457
/                                        35814
Name: page, Length: 1913, dtype: int64

In [61]:
#Look at what page was least viewed for php
php.groupby(['page'])['page'].count().sort_values()

page
html-css/media-queries                                                              1
prework/cli/07-editing-files-with-vim                                               1
content/appendix/javascript/functions/controllers.html                              1
content/appendix/javascript/functions/models.html                                   1
content/appendix/javascript/functions/scope.html                                    1
content/appendix/javascript/functions/templating.html                               1
content/appendix/javascript/javascript/functions/scope.html                         1
content/mysql/relationships/indexes.html                                            1
content/appendix/code-standards/html-css.html                                       1
content/appendix/postwork/mars-rover.html                                           1
content/appendix/postwork/trains.html                                               1
content/conditionals.html                        

## 8. Anything else I should be aware of?

### Why are there so few Front End Graduates accessing the Curriculum?

In [62]:
# Create DataFrame for Front End users
fe_users = df[df.program_id == 'Front End']

In [63]:
# show max end date for Front End users
fe_users.end_date.max()

Timestamp('2015-07-29 00:00:00')

In [64]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [65]:
# How many users and cohorts make up the Front End
fe_users.groupby(['name'])['user_id'].agg(['nunique'])

nunique
name           
Apollo        1

<div class="alert alert-info">


## ANSWER: Because there was only one Front End user who graduated in 2015, three years before the start date of our data access collection which begins in 2018.

## Most likely, the majority of the Null entries for Cohort could probably be attributed to Front End as well as a few towards Staff